<a href="https://colab.research.google.com/github/LelleBSzabo/My-Python-Pandas-Projects/blob/main/pandas1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
!pip install pandas

In [ ]:
import pandas as pd

In [ ]:
pd.__version__

'2.2.2'

# Import Dataset

In [ ]:
df = pd.read_csv("student.csv")

#EDA (Exploratory Data Analysis)

In [ ]:
# first 5 rows
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


In [ ]:
df.head(10)

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
5,6,Alex John,Four,55,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
9,10,Big John,Four,55,female


In [ ]:
df.tail()

,id,name,class,mark,gender
30,31,Marry Toeey,Four,88,male
31,32,Binn Rott,Seven,90,female
32,33,Kenn Rein,Six,96,female
33,34,Gain Toe,Seven,69,male
34,35,Rows Noump,Six,88,female


In [ ]:
df.tail(10)

,id,name,class,mark,gender
25,26,Crelea,Seven,79,male
26,27,NaN,Three,81,NaN
27,28,Rojj Base,Seven,86,female
28,29,Tess Played,Seven,55,male
29,30,Reppy Red,Six,79,female
30,31,Marry Toeey,Four,88,male
31,32,Binn Rott,Seven,90,female
32,33,Kenn Rein,Six,96,female
33,34,Gain Toe,Seven,69,male
34,35,Rows Noump,Six,88,female


In [ ]:
df.sample(7)

,id,name,class,mark,gender
21,22,Reggid,Seven,55,female
14,15,Tade Row,NaN,88,male
2,3,Arnold,Three,55,male
28,29,Tess Played,Seven,55,male
20,21,Babby John,Four,69,female
8,9,Tes Qry,Six,78,NaN
30,31,Marry Toeey,Four,88,male


In [ ]:
df.shape

(35, 5)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 35 entries, 0 to 34
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      35 non-null     int64 
 1   name    34 non-null     object
 2   class   34 non-null     object
 3   mark    35 non-null     int64 
 4   gender  33 non-null     object
dtypes: int64(2), object(3)
memory usage: 1.5+ KB


In [ ]:
df.columns

Index(['id', 'name', 'class', 'mark', 'gender'], dtype='object')

In [ ]:
df.describe()

,id,mark
count,35.000000,35.000000
mean,18.000000,74.657143
std,10.246951,16.401117
min,1.000000,18.000000
25%,9.500000,62.500000
50%,18.000000,79.000000
75%,26.500000,88.000000
max,35.000000,96.000000


*   std = standard deviation
*   25%: 25% of the marks of the students are equal to or less than 62.5
*   50% median. 50% of the marks are equal to or less than 79
* if std is close or greater than mean, there are outliers in the data

In [ ]:
len(df)

35

In [ ]:
df.size

175

In [ ]:
df.dtypes

,0
id,int64
name,object
class,object
mark,int64
gender,object


In [ ]:
df.isnull().sum()

,0
id,0
name,1
class,1
mark,0
gender,2


In [ ]:
df.isnull()

,id,name,class,mark,gender
0,False,False,False,False,False
1,False,False,False,False,False
2,False,False,False,False,False
3,False,False,False,False,False
4,False,False,False,False,False
5,False,False,False,False,False
6,False,False,False,False,False
7,False,False,False,False,False
8,False,False,False,False,True
9,False,False,False,False,False


In [ ]:
df.duplicated().sum()

np.int64(0)

#How to select rows or columns

In [ ]:
df.head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
2,3,Arnold,Three,55,male
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female


##subsetting dataframe

In [ ]:
df["name"]

,name
0,John Deo
1,Max Ruin
2,Arnold
3,Krish Star
4,John Mike
5,Alex John
6,My John Rob
7,Asruid
8,Tes Qry
9,Big John


In [ ]:
df["class"].sample(3)

,class
18,Nine
10,Six
15,Four


In [ ]:
df[["name", "mark"]]

,name,mark
0,John Deo,75
1,Max Ruin,85
2,Arnold,55
3,Krish Star,60
4,John Mike,60
5,Alex John,55
6,My John Rob,78
7,Asruid,85
8,Tes Qry,78
9,Big John,55


In [ ]:
# Write Python code to return name, class, and gender columns

df[["name", "class", "gender"]]

,name,class,gender
0,John Deo,Four,female
1,Max Ruin,Three,male
2,Arnold,Three,male
3,Krish Star,Four,female
4,John Mike,Four,female
5,Alex John,Four,male
6,My John Rob,Fifth,male
7,Asruid,Five,male
8,Tes Qry,Six,NaN
9,Big John,Four,female


In [ ]:
df["mark"][0]

np.int64(75)

#loc and iloc

In [ ]:
df.loc[0:2, ["id", "class", "name"]]

,id,class,name
0,1,Four,John Deo
1,2,Three,Max Ruin
2,3,Three,Arnold


In [ ]:
df.loc[0:4 , ["id", "name", "gender"]]
# these are list of the column names ["id", "name", "gender"]


,id,name,gender
0,1,John Deo,female
1,2,Max Ruin,male
2,3,Arnold,male
3,4,Krish Star,female
4,5,John Mike,female


In [ ]:
df.loc[0:4 , ["id", "name", "gender"]]
# these are list of the column names ["id", "name", "gender"]
# 0:4 is first 5 row
# loc operator includes stopping point of index

,id,name,gender
0,1,John Deo,female
1,2,Max Ruin,male
2,3,Arnold,male
3,4,Krish Star,female
4,5,John Mike,female


In [ ]:
df.iloc[0:5, [0,1,4]]


,id,name,gender
0,1,John Deo,female
1,2,Max Ruin,male
2,3,Arnold,male
3,4,Krish Star,female
4,5,John Mike,female


In [ ]:
# id, class, name of first 3 students
df.loc[0:2, ["id", "class", "name"]]

,id,class,name
0,1,Four,John Deo
1,2,Three,Max Ruin
2,3,Three,Arnold


In [ ]:
df.iloc[0:3, [0,2,1]]

,id,class,name
0,1,Four,John Deo
1,2,Three,Max Ruin
2,3,Three,Arnold


In [ ]:
df.loc[20:30, ["name", "gender"]].shape


(11, 2)

In [ ]:

df[df["mark"]>60]

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
1,2,Max Ruin,Three,85,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
8,9,Tes Qry,Six,78,NaN
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
14,15,Tade Row,NaN,88,male


In [ ]:
df[df["class"]=="Six"]

,id,name,class,mark,gender
8,9,Tes Qry,Six,78,NaN
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
16,17,Tumyu,Six,54,male
29,30,Reppy Red,Six,79,female
32,33,Kenn Rein,Six,96,female
34,35,Rows Noump,Six,88,female


In [ ]:

df[df["gender"]=="female"].head()

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
10,11,Ronald,Six,89,female


In [ ]:
# Female student at class four
# df[(condition1) & (condition)]
df[ (df["gender"]=="female") & (df["class"]=="Four")]

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
20,21,Babby John,Four,69,female


In [ ]:
df.loc[df["gender"]=="female", :].head() # return all columns
# df.loc[rows, columns]

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
3,4,Krish Star,Four,60,female
4,5,John Mike,Four,60,female
9,10,Big John,Four,55,female
10,11,Ronald,Six,89,female


In [ ]:
df.loc[df["gender"]=="female", ["name", "gender"]].head()
# only return name and gender first 5 rows

,name,gender
0,John Deo,female
3,Krish Star,female
4,John Mike,female
9,Big John,female
10,Ronald,female


In [ ]:
df.loc[df["mark"]>90, :]

,id,name,class,mark,gender
11,12,Recky,Six,94,female
32,33,Kenn Rein,Six,96,female


In [ ]:
# df[ (df["mark"]<70) & (df["class"]=="Seven")]
df.loc[(df["mark"]<70) & (df["class"]=="Seven"), :]

,id,name,class,mark,gender
21,22,Reggid,Seven,55,female
28,29,Tess Played,Seven,55,male
33,34,Gain Toe,Seven,69,male


In [ ]:
# Average mark across all students

df["mark"].mean()

np.float64(74.65714285714286)

In [ ]:
# Average mark across male and average mark across female students

df.groupby("gender")["mark"].mean()

,mark
gender,
female,77.312500
male,71.588235


In [ ]:
# Max and min mark

max_mark = df["mark"].max()
min_mark = df["mark"].min()

print("Maximum mark:", max_mark)
print("Minimum mark:", min_mark)

Maximum mark: 96
Minimum mark: 18


In [ ]:
# Max mark for male and max mark for female students

max_mark = df.groupby("gender")["mark"].max()
min_mark = df.groupby("gender")["mark"].min()

print("Maximum mark:", max_mark)
print("Minimum mark:", min_mark)

Maximum mark: gender
female    96
male      88
Name: mark, dtype: int64
Minimum mark: gender
female    55
male      18
Name: mark, dtype: int64


In [ ]:
# Average mark for class 'Five'

df[df["class"]=="Five"]["mark"].mean()

np.float64(80.0)

In [ ]:
# Average mark for class 'Six' and male students only

df[(df["class"] == "Six") & (df["gender"] == "male")]["mark"].mean()

np.float64(54.0)

#Sort Values

In [ ]:
df.sort_values(by="mark").head()

,id,name,class,mark,gender
18,19,Tinny,Nine,18,male
16,17,Tumyu,Six,54,male
2,3,Arnold,Three,55,male
5,6,Alex John,Four,55,male
28,29,Tess Played,Seven,55,male


In [ ]:
df.sort_values(by="name").head()

,id,name,class,mark,gender
5,6,Alex John,Four,55,male
2,3,Arnold,Three,55,male
7,8,Asruid,Five,85,male
20,21,Babby John,Four,69,female
9,10,Big John,Four,55,female


In [ ]:
df["mark"].sort_values().head()

,mark
18,18
16,54
2,55
5,55
28,55


In [ ]:
df.loc[0:4, ["name", "mark"]].sort_values(by="mark", ascending=False)

,name,mark
1,Max Ruin,85
0,John Deo,75
3,Krish Star,60
4,John Mike,60
2,Arnold,55


In [ ]:
# find male student with their mark higher than 70 and sort by their class

df.loc[(df["gender"]=="male") & (df["mark"]>70)].sort_values(by="class")

,id,name,class,mark,gender
22,23,Herod,Eight,79,male
6,7,My John Rob,Fifth,78,male
7,8,Asruid,Five,85,male
17,18,Honny,Five,75,male
15,16,Gimmy,Four,88,male
30,31,Marry Toeey,Four,88,male
23,24,Tiddy Now,Seven,78,male
24,25,Giff Tow,Seven,88,male
25,26,Crelea,Seven,79,male
1,2,Max Ruin,Three,85,male


In [ ]:
df.loc[(df["gender"]=="female") & (df["mark"]>70)]

,id,name,class,mark,gender
0,1,John Deo,Four,75,female
10,11,Ronald,Six,89,female
11,12,Recky,Six,94,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
27,28,Rojj Base,Seven,86,female
29,30,Reppy Red,Six,79,female
31,32,Binn Rott,Seven,90,female
32,33,Kenn Rein,Six,96,female
34,35,Rows Noump,Six,88,female


In [ ]:
#Write code to find information of female students who are above the classroom average. Sort the results in descending order.

df.loc[(df["gender"]=="female") & (df["mark"]>(df["mark"].mean()))].sort_values(by="mark", ascending=False)


,id,name,class,mark,gender
32,33,Kenn Rein,Six,96,female
11,12,Recky,Six,94,female
31,32,Binn Rott,Seven,90,female
10,11,Ronald,Six,89,female
34,35,Rows Noump,Six,88,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
27,28,Rojj Base,Seven,86,female
29,30,Reppy Red,Six,79,female
0,1,John Deo,Four,75,female


In [ ]:
df.loc[(df["gender"]=="female") & (df["mark"]>df["mark"].mean())].sort_values(by="mark", ascending=False)

,id,name,class,mark,gender
32,33,Kenn Rein,Six,96,female
11,12,Recky,Six,94,female
31,32,Binn Rott,Seven,90,female
10,11,Ronald,Six,89,female
34,35,Rows Noump,Six,88,female
12,13,Kty,Seven,88,female
13,14,Bigy,Seven,88,female
27,28,Rojj Base,Seven,86,female
29,30,Reppy Red,Six,79,female
0,1,John Deo,Four,75,female


#Value Counts

In [ ]:
df["gender"].value_counts()

,count
gender,
male,17
female,16


In [ ]:
df["class"].value_counts()

,count
class,
Seven,10
Four,8
Six,7
Three,3
Nine,2
Five,2
Fifth,1
Eight,1


In [ ]:
df["name"].value_counts()

,count
name,
John Deo,1
Max Ruin,1
Arnold,1
Krish Star,1
John Mike,1
Alex John,1
My John Rob,1
Asruid,1
Tes Qry,1
